### PR TITLE
Fix Lightning fallbacks to support PL 0.29 changes

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** PR https://github.com/PennyLaneAI/pennylane/pull/3585 updated how wires were handled for SparseHamiltonian types, where situations of device wires being different to operator wires required a reordering. The assumptions made in Lightning were that operator and device wires were always the same and in the same specific order --- this change broke that assumption. As such, when the two diverge Lightning does not apply the correct values to the appropriate locations.

In addition, due to the changes, fallback functionality in Lightning for non-CPP supported observables werre offloaded to `_qubit_device` with different assumptions made about the bases of said observables. In this case, where _qubit_device required the application of diagonalizing gate, Lightning by default did not, and so the results were inconsistent for said observables. This PR ensure all observable variants for prob, expval and var are offloaded to C++ functions where applicable, and raises an error where functionality support does not currently exist, rather than allowing the return of incorrect results.

**Description of the Change:** Update wire ordering of sparse matrix defined from Hamiltonian to be reflective of matching device and operator order. Ensure appropriate fallbacks are avoided, instead favouring C++ operations. Ensure error raised when attempting out of order wires for probs.

**Benefits:** Fixes edge case where the sparse matrix applied is not correct.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: tests/test_measures.py [hunk 1]
```diff
@@ -116,6 +115,31 @@ def circuit():
 
         assert np.allclose(circuit(), cases[1], atol=tol, rtol=0)
 
+    @pytest.mark.parametrize(
+        "cases",
+        [
+            [[1, 0], [0.9165164490394898, 0.08348355096051052, 0.0, 0.0]],
+            [["a", "0"], [0.9165164490394898, 0.08348355096051052]],
```
Comment (AmintorDusko) [inline tests/test_measures.py:122]:
```suggestion
            [["a", "0"], [0.9165164490394898, 0.08348355096051052, 0.0, 0.0]],
```

**Review by AmintorDusko — APPROVED** [review]
Great work, thanks!
I just made some suggestions, but I'm happy to approve.

File: pennylane_lightning/lightning_qubit.py [hunk 2]
```diff
@@ -744,6 +744,11 @@ def probability(self, wires=None, shot_range=None, bin_size=None):
         Returns:
             array[float]: list of the probabilities
         """
+        if wires and len(wires) > 1 and (not np.all(list(wires)[:-1] <= list(wires)[1:])):
```
Comment (timmysilv) [inline pennylane_lightning/lightning_qubit.py:747]: same comment as in the [kokkos PR](https://github.com/PennyLaneAI/pennylane-lightning-kokkos/pull/41#discussion_r1113490294) - this check should be done on `device_wires` below in case the device itself has a weird wire order

Comment (mlxd) [reply pennylane_lightning/lightning_qubit.py]: Yep, done. Thanks @timmysilv

Comment (mlxd) [reply pennylane_lightning/lightning_qubit.py]: Done.

**Review by timmysilv — APPROVED** [review]
I opened #419 (against this branch) to add more test cases to verify my latest comment. I think they wouldn't hurt to add, but this looks good to me either way. In other words, the tests are there should you want them - take em or leave em.

**Review by vincentmr — APPROVED** [review]
I'm not sure I understand properly what is being compared, so I left a comment about this line

```
(not np.all(list(wires)[:-1] <= list(wires)[1:]))
```

Otherwise, looks good! Thanks.

File: pennylane_lightning/lightning_qubit.py [hunk 3]
```diff
@@ -744,6 +744,11 @@ def probability(self, wires=None, shot_range=None, bin_size=None):
...
+            raise RuntimeError(
+                "Lightning does not currently support out-of-order indices for probabilities"
+            )
+
```
[the remark this answers is not in the document]
Comment (mlxd) [reply pennylane_lightning/lightning_qubit.py]: Ah, good call. Done.
